docs/architecture/adr-054-semver-compatible-modules.md

```diff
@@ -319,24 +319,13 @@ type MsgSend interface {
 func NewMsgSend() MsgSend { return &msgSendImpl{memoryBuffers: ...} }
 ```
 
-Under the hood, `MsgSend` could be implemented based on some raw memory buffers, ex:
-```go
-type msgSendImpl struct {
-	memoryBuffers []interface{}
-}
-
-func (m *msgSendImpl) GetFromAddress() {
-	return m.memoryBuffers[0].(string)
-}
-
-func (m *msgSendImpl) GetToAddress() {
-    return m.memoryBuffers[1].(string)
-}
-```
-
-This approach would have the added benefits of allowing zero-copy message passing to modules written in other languages
-such as Rust. It could also make unknown field filtering in inter-module communication simpler if we require that all
-new fields are added in sequential order, ex. just checking that no field `> 5` is set.
+Under the hood, `MsgSend` could be implemented based on some raw memory buffer in the same way
+that [Cap'n Proto](https://capnproto.org)
+and [FlatBuffers](https://google.github.io/flatbuffers/) so that we could convert between one version of `MsgSend`
+and another without serialization (i.e. zero-copy). This approach would have the added benefits of allowing zero-copy
+message passing to modules written in other languages such as Rust and accessed through a VM or FFI. It could also make
+unknown field filtering in inter-module communication simpler if we require that all new fields are added in sequential
+order, ex. just checking that no field `> 5` is set.
```

Comment:
Language-native constructs like `interface` and `type` and `func` and methods and so on also do not require serialization, and have the additional benefit of incurring no additional costs related to e.g. schema translation or code generation or etc., and require no additional consideration of schema or version management or translation thereof across different semantic domains. Cap'n Proto and FlatBuffers are, like Protocol Buffers or JSON or XML, tools that solve problems of _data serialization_ between process boundaries. They're not meant to express type definitions within a single compilation unit.

Reply:
In general I don't disagree. I think even if we abandoned the idea of treating the inter-module boundary like a process boundary and abandoned the idea of VM modules co-existing, there would still be a few cases of proto versioning leaking into the state machine where we need to deal with the stuff in this ADR. Though maybe that'd be more limited. Like I said, I'll write up an Approach D to explore this.